File: LeetCode/LeetCode 47. 全排列 II (dfs + 全排列).cpp
class Solution {
public:
    int n;
    vector <int> ss;
    vector <vector<int>> ans;

    void dfs(int x, vector<int>& vis, vector<int>& nums) {
        if(x == n) {
            ans.push_back(ss);
            return ;
        }
        for(int i = 0; i < n; i++) {
            if(vis[i] || (i > 0 && nums[i] == nums[i - 1] && !vis[i - 1])) continue;
            vis[i] = 1;
            ss.push_back(nums[i]);
            dfs(x + 1, vis, nums);
            ss.pop_back();
            vis[i] = 0;
        }
        return ;
    }

    vector<vector<int>> permuteUnique(vector<int>& nums) {
        n = nums.size();
        vector <int> vis(n, 0);
        sort(nums.begin(), nums.end());
        dfs(0, vis, nums);
        return ans;
    }
};

File: LeetCode/LeetCode 47. 全排列 II (dfs + 全排列).py
class Solution:
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        n = len(nums)
        vis = [0] * n
        ans = list()
        ss = list()

        def dfs(x: int, nums: List[int]):
            if x == n:
                ans.append(ss[:])
                return 
            for i in range(n):
                if vis[i] == 1 or (i > 0 and nums[i] == nums[i - 1] and vis[i - 1] == 0):
                    continue
                vis[i] = 1
                ss.append(nums[i])
                dfs(x + 1, nums)
                vis[i] = 0
                ss.pop()
            return 
        
        nums.sort()
        dfs(0, nums)
        return ans